gadopt/preconditioners.py

```diff
@@ -0,0 +1,65 @@
+import firedrake as fd
+
+
+class VariableMassInvPC(fd.AuxiliaryOperatorPC):
+    """Scaled inverse pressure mass preconditioner
+
+    Note that unlike MassInvPC this PC will update for variable viscosity."""
+    # use same prefix as MassInvPC, so we can use it as drop-in replacement in options
+    _prefix = 'Mp_'
+
+    def form(self, pc, test, trial):
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+        mu = context.appctx.get("mu", 1.0)
+        # for future Augmented Lagrangian term:
+        gamma = context.appctx.get("gamma", 0.0)
+        a = fd.inner(-1/(mu+gamma) * test, trial) * fd.dx
+        bcs = []
+        return a, bcs
+
+
+class P0MassInvPC(fd.PCBase):
+    """Scaled inverse pressure mass preconditioner to be used with P0 pressure"""
+
+    def initialize(self, pc):
+        _, P = pc.getOperators()
+        appctx = self.get_appctx(pc)
+        V = fd.dmhooks.get_function_space(pc.getDM())
+        # get function spaces
+        assert V.ufl_element().degree() == 0
+        u = fd.TrialFunction(V)
+        v = fd.TestFunction(V)
+        massinv = fd.assemble(fd.Tensor(fd.inner(u, v)*fd.dx).inv)
+        self.massinv = massinv.petscmat
+        self.mu = appctx["mu"]
+        self.gamma = appctx["gamma"]
```

Comment:
Should these be `.get()` like in `VariableMassInvPC`, or maybe they'll always be present anyway?